File: ts/isHappy.ts
function isHappy(n: number): boolean {
  const nMap = new Map<number, boolean>();

  while (n !== 1 && !nMap.has(n)) {
    nMap.set(n, true);
    n = getNextNumber(n);
  }

  return n === 1;
}

function getNextNumber(n: number): number {
  const nextNumber = String(n)
    .split('')
    .reduce((a, b) => {
      return a + Math.pow(Number(b), 2);
    }, 0);

  console.log({ nextNumber });

  return nextNumber;
}

export default () => {
  const n = 2;
  console.log(isHappy(n));
};

File: rs/src/is_happy.rs
use std::collections::HashMap;

fn is_happy(mut n: i32) -> bool {
    let mut seen: HashMap<i32, bool> = HashMap::new();

    while n != 1 && !seen.contains_key(&n) {
        seen.insert(n, true);
        n = get_next_number(n);
    }

    return n.eq(&1);
}

fn get_next_number(n: i32) -> i32 {
    let chars = n.to_string().chars().collect::<Vec<char>>();

    return chars.iter().fold(0, |acc, cur| {
        return acc + cur.to_digit(10).unwrap().pow(2) as i32;
    });
}

pub fn main() {
    let n: i32 = 19;
    let result = is_happy(n);

    println!("{result}");
}
